### PR TITLE
Redesign crates, add Azure support

### PR DESCRIPTION
**Pull request by auguwu**
This crate does a whole lot of re-designing for the long run.

* Removed `remi_core` as `remi` should serve the same purpose
* Restructure arguments to methods that rely on `String`s or `&[u8]` to use an impl of `Into<T>`.
* Having methods "panic" to seal a mut reference is stupid and I wish I never did this! This uses `mut self` instead of `&mut self`.

This crate also adds experimental support for [Azure Blob Storage](https://azure.microsoft.com/en-us/products/storage/blobs) via [`azure_blob_storage`](https://github.com/Azure/azure-sdk-for-rust) crate and [Google Cloud Storage](https://cloud.google.com/storage) via [`google-cloud-storage`](https://github.com/yoshidan/google-cloud-rust/tree/main/storage) crate.

### CODE REVIEW

File: crates/azure/src/service.rs [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+    fn check_if_azurite_container_can_run() {
+        if !is_docker_enabled() {
+            eprintln!("[remi-azure] `docker` is missing, cannot run test");
+            return;
+        }
+
+        // in CI (GitHub Actions), it will pull the `azurite` image as a Windows container and Microsoft
+        // doesn't ship Windows containers of the `azurite` image, so we just ignore the test alltogether
+        //
+        // TODO(@auguwu): how to fix this? :3
```
Comment (spotlightishere) [inline crates/azure/src/service.rs:452]: Would it be possible to split this out into a separate issue?

File: crates/gridfs/src/service.rs [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+        // MongoDB expects strings to be utf-8 encoded?
+        let path_str = path
+            .to_str()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "expected utf-8 encoded path string"))?;
+
+        let mut stream = self.0.open_upload_stream(path_str, None);
+        stream.write_all(&options.data[..]).await?;
+        stream.close().await
+
+        // TODO(@auguwu): add metadata to document that was created and the given content type
```
Comment (spotlightishere) [inline crates/gridfs/src/service.rs:436]: Similarly, can this be split out into its own issue?

**Review by spotlightishere — APPROVED** [review]
This is an extremely large pull request, so I can't quite do an in-depth review! However, from looking at the general scope of changes, this seems like a promising direction - and certainly assists with moving a lot of the AWS S3 SDK specific things out!

I left only quick comments asking to please make `TODO`s in source also become GitHub issues - would appreciate your insight!

**Comment by auguwu** [comment]
> I left only quick comments asking to please make `TODO`s in source also become GitHub issues - would appreciate your insight!

I'll make marked `TODO`s as issues once this PR is merged.

**Comment by auguwu** [comment]
I decided to not work on `remi-gcs`, as I'll plan to do it soon when there is an officially maintained Google Cloud SDK that is reliable or a community-maintained version can be present if wished.

**Pull request merged**